src/MainWindow.vala

```diff
@@ -1421,63 +1412,42 @@ namespace Terminal {
             return (TerminalWidget)((Gtk.Bin)tab.page).get_child ();
         }
 
-        private uint name_check_timeout_id = 0;
-        private void schedule_name_check () {
-            if (name_check_timeout_id > 0) {
-                Source.remove (name_check_timeout_id);
-            }
-
-            name_check_timeout_id = Timeout.add (50, () => {
-                save_opened_terminals ();
-
-                if (!check_for_tabs_with_same_name ()) {
-                    return true;
-                } else {
-                    name_check_timeout_id = 0;
-                    return false;
-                }
-            });
-        }
-
         /** Compare every tab label with every other and resolve ambiguities **/
-        private bool check_for_tabs_with_same_name () {
+        private void check_for_tabs_with_same_name () {
             /* Take list copies so foreach clauses can be nested safely*/
             var terms = terminals.copy ();
             var terms2 = terminals.copy ();
 
             foreach (TerminalWidget terminal in terms) {
-                string term_path = terminal.get_shell_location ();
-                string term_label = terminal.window_title;
+                string term_path = terminal.current_working_directory;
+                string term_label = Path.get_basename (term_path);
 
-                if (term_label == "") { /* No point in continuing - tabs not finished updating */
-                    return false; /* Try again later */
-                }
+                if (term_label == "" ||
+                    terminal.tab_label == TerminalWidget.DEFAULT_LABEL) {
 
-                if (terminal.tab_label == TerminalWidget.DEFAULT_LABEL) { /* Absent or incorrect .bashrc */
-                    return true;
+                    continue;
                 }
 
                 /* Reset tab_name to basename so long name only used when required */
                 terminal.tab_label = term_label;
 
                 foreach (TerminalWidget terminal2 in terms2) {
-                    string term2_path = terminal2.get_shell_location ();
-                    string term2_name = terminal2.window_title;
+                    string term2_path = terminal2.current_working_directory;
+                    string term2_name = Path.get_basename (term2_path);
 
                     if (terminal2 != terminal && term2_name == term_label) {
                         if (term2_path != term_path) {
                             terminal2.tab_label = disambiguate_label (term2_path, term_path);
                             terminal.tab_label = disambiguate_label (term_path, term2_path);
-
-                            if (terminal == current_terminal) {
-                                title = terminal.tab_label;
-                            }
                         }
                     }
                 }
+
+                terminal.tab.tooltip = term_path;
```

Comment:
Would it be possible to move this to `TerminalWidget` and make use of `bind_property` to automatically update the tooltip?

Reply:
Or, just move this to the `check_cwd_changed` function.

Reply:
OK, I'll move it to `check_cwd_changed` it doesn't need to be here.

Reply:
Tried `bind_property` but doesnt work because `Granite.Widgets.Tab` resets the tooltip_text everytime the label text is reassigned - which may happen even when the cwd has not changed.  So I think its better to leave it as it is tbh. We do not want to reassign it on every call to `check_cwd_changed`